File: skills/wechatpay-basic-payment/references/2-服务商/示例代码/Go/17-分账/UnfreezeOrder.go
package main

import (
	"bytes"
	"demo/wxpay_utility" // 引用微信支付工具库，参考 https://pay.weixin.qq.com/doc/v3/partner/4015119446
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"
)

func main() {
	config, err := wxpay_utility.CreateMchConfig(
		"YOUR_MCHID",
		"YOUR_CERT_SERIAL_NO",
		"/path/to/apiclient_key.pem",
		"YOUR_PUB_KEY_ID",
		"/path/to/wxp_pub.pem",
	)
	if err != nil {
		fmt.Println(err)
		return
	}

	request := &UnfreezeOrderRequest{
		SubMchid:      wxpay_utility.String("1900000109"),
		TransactionId: wxpay_utility.String("4208450740201411110007820472"),
		OutOrderNo:    wxpay_utility.String("P20150806125346"),
		Description:   wxpay_utility.String("解冻全部剩余资金"),
	}

	response, err := UnfreezeOrder(config, request)
	if err != nil {
		fmt.Printf("请求失败: %+v\n", err)
		return
	}

	fmt.Printf("请求成功: %+v\n", response)
}

func UnfreezeOrder(config *wxpay_utility.MchConfig, request *UnfreezeOrderRequest) (response *OrdersEntity, err error) {
	const (
		host   = "https://api.mch.weixin.qq.com"
		method = "POST"
		path   = "/v3/profitsharing/orders/unfreeze"
	)

	reqUrl, err := url.Parse(fmt.Sprintf("%s%s", host, path))
	if err != nil {
		return nil, err
	}
	reqBody, err := json.Marshal(request)
	if err != nil {
		return nil, err
	}
	httpRequest, err := http.NewRequest(method, reqUrl.String(), bytes.NewReader(reqBody))
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Accept", "application/json")
	httpRequest.Header.Set("Wechatpay-Serial", config.WechatPayPublicKeyId())
	httpRequest.Header.Set("Content-Type", "application/json")
	authorization, err := wxpay_utility.BuildAuthorization(config.MchId(), config.CertificateSerialNo(), config.PrivateKey(), method, reqUrl.RequestURI(), reqBody)
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Authorization", authorization)

	client := &http.Client{}
	httpResponse, err := client.Do(httpRequest)
	if err != nil {
		return nil, err
	}
	respBody, err := wxpay_utility.ExtractResponseBody(httpResponse)
	if err != nil {
		return nil, err
	}
	if httpResponse.StatusCode >= 200 && httpResponse.StatusCode < 300 {
		err = wxpay_utility.ValidateResponse(config.WechatPayPublicKeyId(), config.WechatPayPublicKey(), &httpResponse.Header, respBody)
		if err != nil {
			return nil, err
		}
		response := &OrdersEntity{}
		if err := json.Unmarshal(respBody, response); err != nil {
			return nil, err
		}
		return response, nil
	} else {
		return nil, wxpay_utility.NewApiException(httpResponse.StatusCode, httpResponse.Header, respBody)
	}
}

type UnfreezeOrderRequest struct {
	SubMchid      *string `json:"sub_mchid,omitempty"`
	TransactionId *string `json:"transaction_id,omitempty"`
	OutOrderNo    *string `json:"out_order_no,omitempty"`
	Description   *string `json:"description,omitempty"`
}

type OrdersEntity struct {
	SubMchid      *string               `json:"sub_mchid,omitempty"`
	TransactionId *string               `json:"transaction_id,omitempty"`
	OutOrderNo    *string               `json:"out_order_no,omitempty"`
	OrderId       *string               `json:"order_id,omitempty"`
	State         *OrderStatus          `json:"state,omitempty"`
	Receivers     []OrderReceiverDetail `json:"receivers,omitempty"`
}

type OrderStatus string

func (e OrderStatus) Ptr() *OrderStatus { return &e }

const (
	ORDERSTATUS_PROCESSING OrderStatus = "PROCESSING"
	ORDERSTATUS_FINISHED   OrderStatus = "FINISHED"
)

type OrderReceiverDetail struct {
	Amount      *int64            `json:"amount,omitempty"`
	Description *string           `json:"description,omitempty"`
	Type        *ReceiverType     `json:"type,omitempty"`
	Account     *string           `json:"account,omitempty"`
	Result      *DetailStatus     `json:"result,omitempty"`
	FailReason  *DetailFailReason `json:"fail_reason,omitempty"`
	CreateTime  *time.Time        `json:"create_time,omitempty"`
	FinishTime  *time.Time        `json:"finish_time,omitempty"`
	DetailId    *string           `json:"detail_id,omitempty"`
}

type ReceiverType string

func (e ReceiverType) Ptr() *ReceiverType { return &e }

const (
	RECEIVERTYPE_MERCHANT_ID         ReceiverType = "MERCHANT_ID"
	RECEIVERTYPE_PERSONAL_OPENID     ReceiverType = "PERSONAL_OPENID"
	RECEIVERTYPE_PERSONAL_SUB_OPENID ReceiverType = "PERSONAL_SUB_OPENID"
)

type DetailStatus string

func (e DetailStatus) Ptr() *DetailStatus { return &e }

const (
	DETAILSTATUS_PENDING DetailStatus = "PENDING"
	DETAILSTATUS_SUCCESS DetailStatus = "SUCCESS"
	DETAILSTATUS_CLOSED  DetailStatus = "CLOSED"
)

type DetailFailReason string

func (e DetailFailReason) Ptr() *DetailFailReason { return &e }

const (
	DETAILFAILREASON_ACCOUNT_ABNORMAL                DetailFailReason = "ACCOUNT_ABNORMAL"
	DETAILFAILREASON_NO_RELATION                     DetailFailReason = "NO_RELATION"
	DETAILFAILREASON_RECEIVER_HIGH_RISK              DetailFailReason = "RECEIVER_HIGH_RISK"
	DETAILFAILREASON_RECEIVER_REAL_NAME_NOT_VERIFIED DetailFailReason = "RECEIVER_REAL_NAME_NOT_VERIFIED"
	DETAILFAILREASON_NO_AUTH                         DetailFailReason = "NO_AUTH"
	DETAILFAILREASON_RECEIVER_RECEIPT_LIMIT          DetailFailReason = "RECEIVER_RECEIPT_LIMIT"
	DETAILFAILREASON_PAYER_ACCOUNT_ABNORMAL          DetailFailReason = "PAYER_ACCOUNT_ABNORMAL"
	DETAILFAILREASON_INVALID_REQUEST                 DetailFailReason = "INVALID_REQUEST"
)

File: skills/wechatpay-basic-payment/references/2-服务商/示例代码/Java/17-分账/UnfreezeOrder.java
package com.java.demo;

import com.java.utils.WXPayUtility; // 引用微信支付工具库，参考：https://pay.weixin.qq.com/doc/v3/partner/4014985777

import com.google.gson.annotations.SerializedName;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.util.ArrayList;
import java.util.List;

/**
 * 解冻剩余资金（服务商模式）
 */
public class UnfreezeOrder {
  private static String HOST = "https://api.mch.weixin.qq.com";
  private static String METHOD = "POST";
  private static String PATH = "/v3/profitsharing/orders/unfreeze";

  public static void main(String[] args) {
    UnfreezeOrder client = new UnfreezeOrder(
      "YOUR_MCHID",
      "YOUR_CERT_SERIAL_NO",
      "/path/to/apiclient_key.pem",
      "YOUR_PUB_KEY_ID",
      "/path/to/wxp_pub.pem"
    );

    UnfreezeOrderRequest request = new UnfreezeOrderRequest();
    request.subMchid = "1900000109";
    request.transactionId = "4208450740201411110007820472";
    request.outOrderNo = "P20150806125346";
    request.description = "解冻全部剩余资金";
    try {
      OrdersEntity response = client.run(request);
      System.out.println(response);
    } catch (WXPayUtility.ApiException e) {
      e.printStackTrace();
    }
  }

  public OrdersEntity run(UnfreezeOrderRequest request) {
    String uri = PATH;
    String reqBody = WXPayUtility.toJson(request);

    Request.Builder reqBuilder = new Request.Builder().url(HOST + uri);
    reqBuilder.addHeader("Accept", "application/json");
    reqBuilder.addHeader("Wechatpay-Serial", wechatPayPublicKeyId);
    reqBuilder.addHeader("Authorization", WXPayUtility.buildAuthorization(mchid, certificateSerialNo, privateKey, METHOD, uri, reqBody));
    reqBuilder.addHeader("Content-Type", "application/json");
    RequestBody requestBody = RequestBody.create(MediaType.parse("application/json; charset=utf-8"), reqBody);
    reqBuilder.method(METHOD, requestBody);
    Request httpRequest = reqBuilder.build();

    OkHttpClient client = new OkHttpClient.Builder().build();
    try (Response httpResponse = client.newCall(httpRequest).execute()) {
      String respBody = WXPayUtility.extractBody(httpResponse);
      if (httpResponse.code() >= 200 && httpResponse.code() < 300) {
        WXPayUtility.validateResponse(this.wechatPayPublicKeyId, this.wechatPayPublicKey,
            httpResponse.headers(), respBody);
        return WXPayUtility.fromJson(respBody, OrdersEntity.class);
      } else {
        throw new WXPayUtility.ApiException(httpResponse.code(), respBody, httpResponse.headers());
      }
    } catch (IOException e) {
      throw new UncheckedIOException("Sending request to " + uri + " failed.", e);
    }
  }

  private final String mchid;
  private final String certificateSerialNo;
  private final PrivateKey privateKey;
  private final String wechatPayPublicKeyId;
  private final PublicKey wechatPayPublicKey;

  public UnfreezeOrder(String mchid, String certificateSerialNo, String privateKeyFilePath, String wechatPayPublicKeyId, String wechatPayPublicKeyFilePath) {
    this.mchid = mchid;
    this.certificateSerialNo = certificateSerialNo;
    this.privateKey = WXPayUtility.loadPrivateKeyFromPath(privateKeyFilePath);
    this.wechatPayPublicKeyId = wechatPayPublicKeyId;
    this.wechatPayPublicKey = WXPayUtility.loadPublicKeyFromPath(wechatPayPublicKeyFilePath);
  }

  public static class UnfreezeOrderRequest {
    @SerializedName("sub_mchid")
    public String subMchid;

    @SerializedName("transaction_id")
    public String transactionId;

    @SerializedName("out_order_no")
    public String outOrderNo;

    @SerializedName("description")
    public String description;
  }

  public static class OrdersEntity {
    @SerializedName("sub_mchid")
    public String subMchid;

    @SerializedName("transaction_id")
    public String transactionId;

    @SerializedName("out_order_no")
    public String outOrderNo;

    @SerializedName("order_id")
    public String orderId;

    @SerializedName("state")
    public OrderStatus state;

    @SerializedName("receivers")
    public List<OrderReceiverDetail> receivers = new ArrayList<OrderReceiverDetail>();
  }

  public enum OrderStatus {
    @SerializedName("PROCESSING")
    PROCESSING,
    @SerializedName("FINISHED")
    FINISHED
  }

  public static class OrderReceiverDetail {
    @SerializedName("amount")
    public Long amount;

    @SerializedName("description")
    public String description;

    @SerializedName("type")
    public ReceiverType type;

    @SerializedName("account")
    public String account;

    @SerializedName("result")
    public DetailStatus result;

    @SerializedName("fail_reason")
    public DetailFailReason failReason;

    @SerializedName("create_time")
    public String createTime;

    @SerializedName("finish_time")
    public String finishTime;

    @SerializedName("detail_id")
    public String detailId;
  }

  public enum ReceiverType {
    @SerializedName("MERCHANT_ID")
    MERCHANT_ID,
    @SerializedName("PERSONAL_OPENID")
    PERSONAL_OPENID,
    @SerializedName("PERSONAL_SUB_OPENID")
    PERSONAL_SUB_OPENID
  }

  public enum DetailStatus {
    @SerializedName("PENDING")
    PENDING,
    @SerializedName("SUCCESS")
    SUCCESS,
    @SerializedName("CLOSED")
    CLOSED
  }

  public enum DetailFailReason {
    @SerializedName("ACCOUNT_ABNORMAL")
    ACCOUNT_ABNORMAL,
    @SerializedName("NO_RELATION")
    NO_RELATION,
    @SerializedName("RECEIVER_HIGH_RISK")
    RECEIVER_HIGH_RISK,
    @SerializedName("RECEIVER_REAL_NAME_NOT_VERIFIED")
    RECEIVER_REAL_NAME_NOT_VERIFIED,
    @SerializedName("NO_AUTH")
    NO_AUTH,
    @SerializedName("RECEIVER_RECEIPT_LIMIT")
    RECEIVER_RECEIPT_LIMIT,
    @SerializedName("PAYER_ACCOUNT_ABNORMAL")
    PAYER_ACCOUNT_ABNORMAL,
    @SerializedName("INVALID_REQUEST")
    INVALID_REQUEST
  }

}
